src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -320,6 +320,20 @@ default MemorySegment asSlice(long offset) {
      */
     boolean isMapped();
 
+    /**
+     * Returns true if this segment overlaps the {@code other} segment.
+     *
+     * If the two segments are not either both backed by native or heap memory, {@code false} is returned.
```

Comment:
If you like the above suggestion, I'd replace this para with:

"As such, it is never possible for a native (javadoc link) segment to overlap with an heap segment".

We can drop the sentence about offset and sizes - as that seems more of an impl detail? But perhaps, if you want to explain more, a couple of examples would be better? E.g.

```
MemorySegment s1 = MemorySegment.allocateNative(100, scope);
MemorySegment s2 = MemorySegment.allocateNative(100, scope);
s1.overlap(s2); // false
MemorySegment s3 = s1.slice(0, 50);
s1.overlap(slice); // true
s2.overlap(slice); // false
```